### PR TITLE
[cdc-connector][db2] Db2 support incremental source

### LINKED ISSUE #2869: [cdc-connector][db2] Db2 support incremental source
### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.

### Motivation

db2 support incremental source

### Solution

Db2 debezium is inspired by sqlserver debezium, So db2 incremental source will refer to the implementation of SQL Server CDC incremental source. 

 

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by gong**
### Prepare a Pull Request

-  Fixes #2869 

### Motivation

Db2 support incremental source

### Modifications

* Add db2 incremental source code
* Add doc
* Add ITcase UT and e2e UT

### CODE REVIEW

**Comment by gong** [comment]
@leonardBang @GOODBOY008 PTAL

**Comment by gong** [comment]
@leonardBang core UT fail. because this PR https://github.com/ververica/flink-cdc-connectors/pull/2845 cause error. I rollback this change.

**Comment by gong** [comment]
random UT fail that is not relate with the db2

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/table/StartupOptions.java [hunk 1]
```diff
@@ -1,70 +0,0 @@
...
- * limitations under the License.
- */
-
-package com.ververica.cdc.connectors.db2.table;
-
-import java.util.Objects;
-
-/** Debezium startup options. */
-public final class StartupOptions {
-    public final StartupMode startupMode;
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/table/StartupOptions.java:23]: > `StartupMode` class can be removed too.

@lvyanquan done.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2TypeUtils.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            case Types.SMALLINT:
+                return DataTypes.INT();
+            case Types.BIGINT:
+                return DataTypes.BIGINT();
+            case Types.FLOAT:
+            case Types.REAL:
+            case Types.DOUBLE:
+            case Types.DECIMAL:
+            case Types.NUMERIC:
+                return DataTypes.DECIMAL(column.length(), column.scale().orElse(0));
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2TypeUtils.java:64]: What is this for converting FLOAT/DOUBLE to DataTypes.DECIMAL? There is no special handling in [Db2ValueConverters](https://github.com/debezium/debezium-connector-db2/blob/86b3f156f5e7544c5f746387e4749025b5685e0e/src/main/java/io/debezium/connector/db2/Db2ValueConverters.java#L61).

**Review by lvyanquan — COMMENTED** [review]
Thanks for this great contribution, left some comments.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+    public static long queryApproximateRowCnt(JdbcConnection jdbc, TableId tableId)
+            throws SQLException {
+        // The statement used to get approximate row count which is less
+        // accurate than COUNT(*), but is more efficient for large table.
+        final String rowCountQuery =
+                String.format(
+                        "SELECT CARD\n"
+                                + "FROM SYSCAT.TABLES\n"
+                                + "where TABSCHEMA='%s' and TABNAME='%s';",
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java:86]: It is recommended to maintain consistent case sensitivity in SQL.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+            String splitColumnName,
+            int chunkSize,
+            Object includedLowerBound)
+            throws SQLException {
+        String quotedColumn = quote(splitColumnName);
+        String query =
+                String.format(
+                        "SELECT MAX(%s) FROM ("
+                                + "SELECT %s FROM %s WHERE %s >= ? ORDER BY %s ASC "
+                                + " fetch first %s rows only ) AS T",
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java:138]: Maintain consistent case sensitivity in SQL too.

Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java]: done

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java (hunk 3 above)
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2Utils.java]: done

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2TypeUtils.java (hunk 2 above)
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/utils/Db2TypeUtils.java:68]: I modify it. Thank you.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/dialect/Db2ChunkSplitter.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+    @Override
+    public Collection<SnapshotSplit> generateSplits(TableId tableId) {
+        try (JdbcConnection jdbc = dialect.openJdbcConnection(sourceConfig)) {
+
+            LOG.info("Start splitting table {} into chunks...", tableId);
+            long start = System.currentTimeMillis();
+
+            Table table = dialect.queryTableSchema(jdbc, tableId).getTable();
+            Column splitColumn = Db2Utils.getSplitColumn(table);
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/dialect/Db2ChunkSplitter.java]: I add supportting for `scan.incremental.snapshot.chunk.key-column` option.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/Db2SourceBuilder.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+    /** Password to use when connecting to the Db2 database server. */
+    public Db2SourceBuilder<T> password(String password) {
+        this.configFactory.password(password);
+        return this;
+    }
+
+    /**
+     * The session time zone in database server, e.g. "America/Los_Angeles". It controls how the
+     * TIMESTAMP type in Db2 converted to STRING. See more
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/Db2SourceBuilder.java:93]: done

File: docs/content/connectors/db2-cdc.md [hunk 7]
```diff
@@ -174,6 +166,38 @@ for more detailed information.</td>
...
+    </tr> 
+    <tr>
+      <td>chunk-key.even-distribution.factor.upper-bound</td>
+      <td>optional</td>
+      <td style="word-wrap: break-word;">1000.0d</td>
+      <td>Double</td>
+      <td>The upper bound of chunk key distribution factor. The distribution factor is used to determine whether the table is evenly distribution or not. 
+          The table chunks would use evenly calculation optimization when the data distribution is even, and the query for splitting would happen when it is uneven. 
+          The distribution factor could be calculated by (MAX(id) - MIN(id) + 1) / rowCount.</td>
+    </tr>
```
[the remark this answers is not in the document]
Comment (gong) [reply docs/content/connectors/db2-cdc.md:200]: done

File: docs/content/connectors/db2-cdc.md [hunk 8]
```diff
@@ -184,6 +216,17 @@ Db2 server.
...
+      <td>scan.incremental.close-idle-reader.enabled</td>
+      <td>optional</td>
+      <td style="word-wrap: break-word;">false</td>
+      <td>Boolean</td>
+      <td>Whether to close idle readers at the end of the snapshot phase. <br>
+          The flink version is required to be greater than or equal to 1.14 when 'execution.checkpointing.checkpoints-after-tasks-finish.enabled' is set to true.<br>
+          If the flink version is greater than or equal to 1.15, the default value of 'execution.checkpointing.checkpoints-after-tasks-finish.enabled' has been changed to true,
+          so it does not need to be explicitly configured 'execution.checkpointing.checkpoints-after-tasks-finish.enabled' = 'true'
+      </td>
+    </tr>
```
Comment (lvyanquan) [inline docs/content/connectors/db2-cdc.md:229]: Please help to add a `Available Metadata` section, like [Available Metadata](https://ververica.github.io/flink-cdc-connectors/master/content/connectors/sqlserver-cdc.html#available-metadata).

**Review by lvyanquan — COMMENTED** [review]
Overall look goo to me, left some comments.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/test/java/com/ververica/cdc/connectors/db2/table/Db2ConnectorITCase.java [hunk 9]
```diff
@@ -208,16 +206,14 @@ public void testAllTypes() throws Exception {
                                 + " 'username' = '%s',"
                                 + " 'password' = '%s',"
                                 + " 'database-name' = '%s',"
-                                + " 'schema-name' = '%s',"
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/test/java/com/ververica/cdc/connectors/db2/table/Db2ConnectorITCase.java:211]: done

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/fetch/Db2ScanFetchTask.java [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+public class Db2ScanFetchTask extends AbstractScanFetchTask {
+
+    public Db2ScanFetchTask(SnapshotSplit split) {
+        super(split);
+    }
+
+    @Override
+    protected void executeDataSnapshot(Context context) throws Exception {
+        Db2SourceFetchTaskContext sourceFetchContext = (Db2SourceFetchTaskContext) context;
+        taskRunning = true;
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/fetch/Db2ScanFetchTask.java:68]: done

File: docs/content/connectors/db2-cdc.md (hunk 8 above)
Comment (gong) [reply docs/content/connectors/db2-cdc.md:229]: done

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/fetch/Db2ScanFetchTask.java [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+        protected SnapshotResult<Db2OffsetContext> doExecute(
+                ChangeEventSourceContext context,
+                Db2OffsetContext previousOffset,
+                SnapshotContext<Db2Partition, Db2OffsetContext> snapshotContext,
+                SnapshottingTask snapshottingTask)
+                throws Exception {
+            final Db2SnapshotContext ctx = (Db2SnapshotContext) snapshotContext;
+            ctx.offset = offsetContext;
+
+            LOG.info("Snapshot step 2 - Snapshotting data");
```
[the remark this answers is not in the document]
Comment (gong) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-db2-cdc/src/main/java/com/ververica/cdc/connectors/db2/source/fetch/Db2ScanFetchTask.java:214]: done

**Comment by lvyanquan** [comment]
Hi @gong  There is a failure in https://dev.azure.com/ververica/flink-cdc-connectors/_build/results?buildId=4654&view=logs&j=450e5386-ec4e-5c3c-8d86-ea427d28cb6f&t=b72b380c-c1b9-55cc-a168-123a50838bbf, could you please take a look?

**Comment by gong** [comment]
> Hi @gong There is a failure in https://dev.azure.com/ververica/flink-cdc-connectors/_build/results?buildId=4654&view=logs&j=450e5386-ec4e-5c3c-8d86-ea427d28cb6f&t=b72b380c-c1b9-55cc-a168-123a50838bbf, could you please take a look?

ok

**Review by lvyanquan — APPROVED** [review]
LGTM. 
There is a mix of `binlog` `redolog` and `wal` in the code, it is best to unify them.

**Comment by gong** [comment]
> LGTM. There is a mix of `binlog` `redolog` and `wal` in the code, it is best to unify them.

ok

**Comment by gong** [comment]
I will rebase code after https://github.com/ververica/flink-cdc-connectors/pull/2978 is merged.

**Comment by gong** [comment]
@lvyanquan CI error is not related with the PR. UT can pass in local.

**Comment by leonardBang** [comment]
Thanks @gong for the great work, I'd like to have a final review base on @lvyanquan 's review work.

**Review by leonardBang — APPROVED** [review]
Thanks @gong for the contribution and @lvyanquan for the review,  generally looks good to me, I append a minor commit to improve the code style.

**Comment by leonardBang** [comment]
CI passed, merging...

**Pull request merged**